includes/utils/class-amp-image-dimension-extractor.php

```diff
@@ -150,6 +151,128 @@ private static function register_callbacks() {
 		do_action( 'amp_extract_image_dimensions_batch_callbacks_registered' );
 	}
 
+	/**
+	 * To get attachment ID from attached path.
+	 *
+	 * @param string $path Attached path.
+	 *
+	 * @return int Positive number on success, Otherwise 0.
+	 */
+	private static function get_attachment_id_from_path( $path ) {
+
+		if ( empty( $path ) ) {
+			return 0;
+		}
+
+		$path = wp_parse_url( $path, PHP_URL_PATH );
+
+		if ( empty( $path ) ) {
+			return 0;
+		}
+
+		global $wpdb;
+
+		$cache_key     = md5( $path );
+		$cache_group   = 'amp_attachment_id_from_path';
+		$attachment_id = wp_cache_get( $cache_key, $cache_group );
+
+		if ( empty( $attachment_id ) ) {
+			$attachment_id = $wpdb->get_col( // phpcs:ignore WordPress.DB.DirectDatabaseQuery.DirectQuery
+				$wpdb->prepare(
+					"SELECT post_id FROM $wpdb->postmeta WHERE meta_key='_wp_attached_file' AND meta_value=%s ORDER BY post_id ASC;",
+					$path
+				)
+			);
+
+			$attachment_id = ( ! empty( $attachment_id ) && is_array( $attachment_id ) ) ? array_pop( $attachment_id ) : 0;
+
+			wp_cache_set( $cache_key, $attachment_id, $cache_group, 3 * HOUR_IN_SECONDS );
+		}
+
+		return (int) $attachment_id;
+	}
+
+	/**
+	 * Extract dimensions from filename if dimension exists or from file system.
+	 *
+	 * @param array $dimensions Image urls mapped to dimensions.
+	 *
+	 * @return array Dimensions mapped to image urls, or false if they could not be retrieved
+	 */
+	public static function extract_by_filename_or_filesystem( $dimensions ) {
+
+		if ( empty( $dimensions ) || ! is_array( $dimensions ) ) {
+			return [];
+		}
+
+		$upload_dir      = wp_get_upload_dir();
+		$base_upload_uri = strtolower( trim( $upload_dir['baseurl'] ) );
+
+		foreach ( $dimensions as $url => $value ) {
+
+			// Check whether some other callback attached to the filter already provided dimensions for this image.
+			if ( ! empty( $value ) && is_array( $value ) ) {
+				continue;
+			}
+
+			// Check if it's internal media or not. If it's not then bail out.
+			if ( false === strpos( strtolower( trim( $url ) ), $base_upload_uri ) ) {
```

Comment:
Why `strtolower`?

Reply:
We'd want to make sure the string appears at the beginning:

```suggestion
			if ( 0 !== strpos( strtolower( trim( $url ) ), $base_upload_uri ) ) {
```

Reply:
Addressed.